llama_index/llms/custom.py

```diff
@@ -0,0 +1,54 @@
+from typing import Any, Sequence
+
+from llama_index.llms.base import (
+    LLM,
+    ChatMessage,
+    ChatResponse,
+    ChatResponseGen,
+    CompletionResponse,
+    CompletionResponseGen,
+)
+from llama_index.llms.generic_utils import (
+    completion_to_chat_decorator,
+    stream_completion_to_chat_decorator,
+)
+
+
+class CustomLLM(LLM):
+    """Simple abstract base class for custom LLMs.
+
+    Subclasses must implement the `__init__`, `complete`,
+        `stream_complete`, and `metadata` methods.
+    """
+
+    def chat(self, messages: Sequence[ChatMessage], **kwargs: Any) -> ChatResponse:
+        chat_fn = completion_to_chat_decorator(self.complete)
+        return chat_fn(messages, **kwargs)
+
+    def stream_chat(
+        self, messages: Sequence[ChatMessage], **kwargs: Any
+    ) -> ChatResponseGen:
+        stream_chat_fn = stream_completion_to_chat_decorator(self.stream_complete)
+        return stream_chat_fn(messages, **kwargs)
+
+    async def achat(
+        self,
+        messages: Sequence[ChatMessage],
+        **kwargs: Any,
+    ) -> ChatResponse:
+        return self.chat(messages, **kwargs)
```

Comment:
I'm just wondering if in a complex pipeline stiching together many local models and remote models, async would be desired.

For this use case, spawning a subprocess to host e.g. the huggingface model, and treat it as an API service would be desired. Not sure if there is an abstraction within langchain to do so.

Reply:
yep, async is definitely desired here. The `CustomLLM` class is mostly trying to provide a minimal interface to hook up a custom LLM model into the framework, ignoring proper async handling.

Reply:
spinning up a local local model in a separate process makes sense to me. Don't think there's something out of the box to support that right now yet. Would be cool to have this in the future for sure!